src/main/java/com/google/devtools/build/lib/remote/RemoteExecutionService.java

```diff
@@ -1228,13 +1229,15 @@ public InMemoryOutput downloadOutputs(RemoteAction action, RemoteActionResult re
     }
 
     FileOutErr outErr = action.getSpawnExecutionContext().getFileOutErr();
-
-    // Always download the action stdout/stderr.
     FileOutErr tmpOutErr = outErr.childOutErr();
-    List<ListenableFuture<Void>> outErrDownloads =
-        remoteCache.downloadOutErr(context, result.actionResult, tmpOutErr);
-    for (ListenableFuture<Void> future : outErrDownloads) {
-      downloadsBuilder.add(transform(future, (v) -> null, directExecutor()));
+
+    // Skip downloading the action stdout/stderr when remote_download_outputs=minimal and the action is successful.
+    if (remoteOptions.remoteOutputsMode != RemoteOutputsMode.MINIMAL || result.getExitCode() != 0) {
+      List<ListenableFuture<Void>> outErrDownloads =
+          remoteCache.downloadOutErr(context, result.actionResult, tmpOutErr);
+      for (ListenableFuture<Void> future : outErrDownloads) {
+        downloadsBuilder.add(transform(future, (v) -> null, directExecutor()));
+      }
```

Comment:
Even if we don't download them eagerly, we might still want to arrange for these files to be browsable through an output service and/or explicitly downloaded via `--remote_download_regex`.

However, see also my top-level comment - I think we should make a decision on whether gating this by the `RemoteOutputsMode` is the right approach, before we put more effort down this path.

Reply:
Right now i assume we are targeting toplevel download mode or FUSE for local development UX, where user might have a need to inspect these locally. That left minimal being the ideal for CI/automation setup, where most of these are already available through a BES/RBE service.

Reply:
Giving this a bit more thought: 

- `--ui_event_filters` and `--output_filter` might work, but they are mostly used to control how the outputs are "displayed" and not "downloaded". In some setups, users might want to have separate control knobs for these. (i.e. download the logs but don't display them for some post-run analysis process)

- We can reuse `shouldDownload(result, execPath)` so `--remote_download_regex` is applied as well as covering top_level. Im just not sure if that's the right UX. I will play with it a bit today to see how it goes.

Reply:
https://github.com/bazelbuild/bazel/commit/3b3a57795ab8292626aeedccc1f4e3f4f494c0b3

Here is a quick draft using the `shouldDownload` approach. I wrote up my findings in the commit's message.

A not-so-obvious thing I missed initially was Bazel does not retain the stdout/stderr files. Instead, it just downloads these into some temp files to dump them out into the terminal and delete them afterward. This means the temp files Path/PathFragment were not eligible to feed into `shouldDownload()` because it uses a different `FileSystem` than the `RemoteOutputChecker`. I managed to piggyback onto the download of output files/dirs to conditionally download the stderr/out, but there are some interesting test failures.

I made the code change in a separate branch because I am not convinced that's the right approach just yet.

I will poke at `--ui_event_filters` approach today. I suspect in a hybrid build where actions are half local, half remote, users might still want to get stderr/out there. But perhaps adding a new value into these flags (i.e. `-remote_stdout,-remote_stderr`) could work. (?)

Reply:
ah ok, so the shouldDownload approach has issue with `experimental_split_xml_generation` being enabled by default. This checks for stdout existence

https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/exec/StandaloneTestStrategy.java;l=834;drc=300c5867b7d2da1ba32abc20e95662096c2a7a08

and depends on it. Hmm 🤔

Reply:
Digged a bit more and I found that there are other special interactions for tests when doing none-minimal download mode. For example https://cs.opensource.google/bazel/bazel/+/887e1bd1ad860da91226ba67dfca0eebd8754e8c:src/main/java/com/google/devtools/build/lib/remote/RemoteOutputChecker.java;l=188-191

So I would rather keep this PR at `minimal` mode for now.